File: bindings/java/src/test/java/org/indunet/robot/bus/interop/InteropPeer.java
package org.indunet.robot.bus.interop;

import java.util.concurrent.atomic.AtomicReference;
import org.indunet.robot.bus.Node;
import org.indunet.robot.bus.NodeOptions;
import org.indunet.robot.bus.sensor_msgs.msg.v1.Imu;
import org.indunet.robot.bus.std_srvs.srv.v1.SetBoolRequest;
import org.indunet.robot.bus.std_srvs.srv.v1.SetBoolResponse;

/**
 * Env-driven interop peer for {@code tests/interop/}.
 *
 * <p>Roles: {@code sub} (Imu), {@code svc-server} (SetBool).
 */
public final class InteropPeer {
    private static final String TOPIC = "/interop/imu";
    private static final String SERVICE = "/interop/set_bool";
    private static final double EXPECT_Z = 0.42;

    private InteropPeer() {}

    public static void main(String[] args) throws Exception {
        String role = requireEnv("ROBOT_BUS_INTEROP_ROLE");
        NodeOptions opts = optionsFromEnv();
        switch (role) {
            case "sub":
                runSub(opts);
                break;
            case "svc-server":
                runSvcServer(opts);
                break;
            default:
                throw new IllegalArgumentException("unknown ROBOT_BUS_INTEROP_ROLE: " + role);
        }
    }

    private static void runSub(NodeOptions opts) throws Exception {
        AtomicReference<Imu> got = new AtomicReference<>();
        try (Node node = new Node("interop_java_sub", opts)) {
            node.createSubscription(TOPIC, (topic, msg) -> got.set(msg), Imu.class);
            node.start();
            System.out.println("READY");
            long deadline = System.currentTimeMillis() + 8000;
            while (got.get() == null && System.currentTimeMillis() < deadline) {
                Thread.sleep(20);
            }
            Imu imu = got.get();
            if (imu == null) {
                throw new IllegalStateException("timed out waiting for Imu on " + TOPIC);
            }
            double z = imu.getAngularVelocity().getZ();
            if (Math.abs(z - EXPECT_Z) > 1e-9) {
                throw new IllegalStateException("unexpected z=" + z);
            }
            node.shutdown();
            node.waitForShutdown();
        }
    }

    private static void runSvcServer(NodeOptions opts) throws Exception {
        try (Node node = new Node("interop_java_svc", opts)) {
            node.createService(
                    SERVICE,
                    req ->
                            SetBoolResponse.newBuilder()
                                    .setSuccess(true)
                                    .setMessage("set:" + (req.getData() ? "true" : "false"))
                                    .build(),
                    SetBoolRequest.class,
                    SetBoolResponse.class);
            node.start();
            System.out.println("READY");
            Thread.sleep(15_000);
            node.shutdown();
            node.waitForShutdown();
        }
    }

    private static NodeOptions optionsFromEnv() {
        return new NodeOptions(
                null,
                "tcp",
                null,
                requireEnv("ROBOT_BUS_MESSAGE_XSUB"),
                requireEnv("ROBOT_BUS_MESSAGE_XPUB"),
                requireEnv("ROBOT_BUS_SERVICE_FRONTEND"),
                requireEnv("ROBOT_BUS_SERVICE_BACKEND"),
                requireEnv("ROBOT_BUS_ACTION_BACKEND"),
                requireEnv("ROBOT_BUS_ACTION_FRONTEND"));
    }

    private static String requireEnv(String key) {
        String v = System.getenv(key);
        if (v == null || v.isEmpty()) {
            throw new IllegalStateException("missing env " + key);
        }
        return v;
    }
}

File: bindings/cpp/tests/interop_peer.cpp
// Cross-language interop peer (env-driven endpoints).
// Roles: svc-server | act-client
#include "harness.hpp"

#include <robot_bus/example_interfaces/action/v1/fibonacci.pb.h>
#include <robot_bus/std_srvs/srv/v1/set_bool.pb.h>

#include <chrono>
#include <cstdlib>
#include <iostream>
#include <string>
#include <thread>
#include <utility>
#include <vector>

namespace {

constexpr const char *kService = "/interop/set_bool";
constexpr const char *kAction = "/interop/fibonacci";

std::string require_env(const char *key) {
  const char *v = std::getenv(key);
  if (!v || !*v) {
    std::cerr << "missing env " << key << "\n";
    std::exit(1);
  }
  return std::string(v);
}

struct EndpointBag {
  std::string message_xsub;
  std::string message_xpub;
  std::string service_frontend;
  std::string service_backend;
  std::string action_frontend;
  std::string action_backend;

  static EndpointBag from_env() {
    EndpointBag e;
    e.message_xsub = require_env("ROBOT_BUS_MESSAGE_XSUB");
    e.message_xpub = require_env("ROBOT_BUS_MESSAGE_XPUB");
    e.service_frontend = require_env("ROBOT_BUS_SERVICE_FRONTEND");
    e.service_backend = require_env("ROBOT_BUS_SERVICE_BACKEND");
    e.action_frontend = require_env("ROBOT_BUS_ACTION_FRONTEND");
    e.action_backend = require_env("ROBOT_BUS_ACTION_BACKEND");
    return e;
  }

  RobotBusNodeOptions to_opts() const {
    RobotBusNodeOptions opts{};
    opts.host = nullptr;
    opts.transport = nullptr;
    opts.ws_url = nullptr;
    opts.message_xsub = message_xsub.c_str();
    opts.message_xpub = message_xpub.c_str();
    opts.service_frontend = service_frontend.c_str();
    opts.service_backend = service_backend.c_str();
    opts.action_frontend = action_frontend.c_str();
    opts.action_backend = action_backend.c_str();
    return opts;
  }
};

void run_svc_server() {
  auto ends = EndpointBag::from_env();
  auto opts = ends.to_opts();
  robot_bus::Node node("interop_cpp_svc", opts);
  auto svc = node.create_service(kService, [](robot_bus::BytesView body) {
    std_srvs::srv::v1::SetBoolRequest req;
    ROBOT_BUS_CHECK(req.ParseFromArray(body.data, static_cast<int>(body.size)));
    std_srvs::srv::v1::SetBoolResponse resp;
    resp.set_success(true);
    resp.set_message(std::string("set:") + (req.data() ? "true" : "false"));
    std::string out;
    ROBOT_BUS_CHECK(resp.SerializeToString(&out));
    return std::vector<uint8_t>(out.begin(), out.end());
  });
  node.start();
  std::cout << "READY" << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(15));
  node.shutdown();
  node.wait();
  (void)svc;
}

void run_act_client() {
  std::this_thread::sleep_for(std::chrono::milliseconds(400));
  auto ends = EndpointBag::from_env();
  auto opts = ends.to_opts();
  robot_bus::Node node("interop_cpp_act_client", opts);
  auto client = node.create_action_client(kAction);

  example_interfaces::action::v1::FibonacciGoal goal;
  goal.set_order(5);
  std::string goal_bytes;
  ROBOT_BUS_CHECK(goal.SerializeToString(&goal_bytes));

  auto handle = client.send_goal(goal_bytes, {}, nullptr, 10.0);
  auto event = handle.wait_result(10.0);
  ROBOT_BUS_CHECK(event.kind == "RESULT");
  example_interfaces::action::v1::FibonacciResult result;
  ROBOT_BUS_CHECK(
      result.ParseFromArray(event.body.data(), static_cast<int>(event.body.size())));
  ROBOT_BUS_CHECK(result.sequence_size() == 5);
  ROBOT_BUS_CHECK(result.sequence(0) == 0);
  ROBOT_BUS_CHECK(result.sequence(1) == 1);
  ROBOT_BUS_CHECK(result.sequence(2) == 1);
  ROBOT_BUS_CHECK(result.sequence(3) == 2);
  ROBOT_BUS_CHECK(result.sequence(4) == 3);
  std::cout << "READY" << std::endl;
}

}  // namespace

int main() {
  const char *role = std::getenv("ROBOT_BUS_INTEROP_ROLE");
  if (!role) {
    std::cerr << "ROBOT_BUS_INTEROP_ROLE required\n";
    return 1;
  }
  const std::string r(role);
  if (r == "svc-server") {
    run_svc_server();
  } else if (r == "act-client") {
    run_act_client();
  } else {
    std::cerr << "unknown role: " << r << "\n";
    return 1;
  }
  return 0;
}
